File: frontend/src/api/boards.ts
import type { Board, Column } from "../types";
import client from "./client";

export interface BoardCreate {
  name: string;
  description?: string;
}

export interface BoardUpdate {
  name?: string;
  description?: string;
}

export interface ColumnCreate {
  name: string;
  position: number;
  is_done_column?: boolean;
}

export interface ColumnUpdate {
  name?: string;
  is_done_column?: boolean;
}

export interface ColumnReorderItem {
  id: string;
  position: number;
}

export const getBoards = (): Promise<Board[]> =>
  client.get<Board[]>("/boards").then((r) => r.data);

export const getBoard = (boardId: string): Promise<Board> =>
  client.get<Board>(`/boards/${boardId}`).then((r) => r.data);

export const createBoard = (payload: BoardCreate): Promise<Board> =>
  client.post<Board>("/boards", payload).then((r) => r.data);

export const updateBoard = (boardId: string, payload: BoardUpdate): Promise<Board> =>
  client.put<Board>(`/boards/${boardId}`, payload).then((r) => r.data);

export const deleteBoard = (boardId: string): Promise<void> =>
  client.delete(`/boards/${boardId}`).then(() => undefined);

export const getColumns = (boardId: string): Promise<Column[]> =>
  client.get<Column[]>(`/boards/${boardId}/columns`).then((r) => r.data);

export const createColumn = (boardId: string, payload: ColumnCreate): Promise<Column> =>
  client.post<Column>(`/boards/${boardId}/columns`, payload).then((r) => r.data);

export const updateColumn = (columnId: string, payload: ColumnUpdate): Promise<Column> =>
  client.put<Column>(`/columns/${columnId}`, payload).then((r) => r.data);

export const reorderColumns = (boardId: string, columns: ColumnReorderItem[]): Promise<Column[]> =>
  client.patch<Column[]>(`/boards/${boardId}/columns/reorder`, { columns }).then((r) => r.data);

export const deleteColumn = (columnId: string): Promise<void> =>
  client.delete(`/columns/${columnId}`).then(() => undefined);

File: backend/app/routers/boards.py
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.models.board import Board
from app.models.column import Column
from app.schemas.board import BoardCreate, BoardResponse, BoardUpdate

router = APIRouter(tags=["boards"])

DEFAULT_COLUMNS = [
    {"name": "To Do", "position": 0, "is_done_column": False},
    {"name": "In Progress", "position": 1, "is_done_column": False},
    {"name": "Review", "position": 2, "is_done_column": False},
    {"name": "Done", "position": 3, "is_done_column": True},
]


async def _get_board_with_columns(board_id: str, db: AsyncSession) -> Board:
    result = await db.execute(
        select(Board)
        .options(selectinload(Board.columns))
        .where(Board.id == board_id)
    )
    board = result.scalar_one_or_none()
    if not board:
        raise HTTPException(status_code=404, detail=f"Board {board_id} not found")
    return board


@router.get("/boards", response_model=list[BoardResponse])
async def list_boards(db: AsyncSession = Depends(get_db)):
    result = await db.execute(
        select(Board).options(selectinload(Board.columns))
    )
    return result.scalars().all()


@router.post("/boards", response_model=BoardResponse, status_code=201)
async def create_board(payload: BoardCreate, db: AsyncSession = Depends(get_db)):
    board = Board(id=str(uuid.uuid4()), name=payload.name, description=payload.description)
    db.add(board)
    await db.flush()

    for col in DEFAULT_COLUMNS:
        db.add(Column(
            id=str(uuid.uuid4()),
            board_id=board.id,
            name=col["name"],
            position=col["position"],
            is_done_column=col["is_done_column"],
        ))

    await db.commit()
    await db.refresh(board)
    return await _get_board_with_columns(board.id, db)


@router.get("/boards/{board_id}", response_model=BoardResponse)
async def get_board(board_id: str, db: AsyncSession = Depends(get_db)):
    return await _get_board_with_columns(board_id, db)


@router.put("/boards/{board_id}", response_model=BoardResponse)
async def update_board(board_id: str, payload: BoardUpdate, db: AsyncSession = Depends(get_db)):
    board = await _get_board_with_columns(board_id, db)
    if payload.name is not None:
        board.name = payload.name
    if payload.description is not None:
        board.description = payload.description
    await db.commit()
    await db.refresh(board)
    return await _get_board_with_columns(board.id, db)


@router.delete("/boards/{board_id}", status_code=204)
async def delete_board(board_id: str, db: AsyncSession = Depends(get_db)):
    board = await _get_board_with_columns(board_id, db)
    await db.delete(board)
    await db.commit()
